server/src/main/java/org/elasticsearch/cluster/coordination/PreVoteCollector.java

```diff
@@ -0,0 +1,192 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cluster.coordination;
+
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.coordination.CoordinationState.VoteCollection;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.component.AbstractComponent;
+import org.elasticsearch.common.lease.Releasable;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.threadpool.ThreadPool.Names;
+import org.elasticsearch.transport.TransportException;
+import org.elasticsearch.transport.TransportResponseHandler;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.elasticsearch.cluster.coordination.CoordinationState.isElectionQuorum;
+import static org.elasticsearch.common.util.concurrent.ConcurrentCollections.newConcurrentSet;
+
+public class PreVoteCollector extends AbstractComponent {
+
+    public static final String REQUEST_PRE_VOTE_ACTION_NAME = "internal:cluster/request_pre_vote";
+
+    private final TransportService transportService;
+    private final Runnable startElection;
+
+    @Nullable
+    private volatile DiscoveryNode currentLeader; // null if there is currently no known leader
+    private volatile PreVoteResponse preVoteResponse;
+
+    PreVoteCollector(final Settings settings, final PreVoteResponse preVoteResponse,
+                     final TransportService transportService, final Runnable startElection) {
+        super(settings);
+        this.preVoteResponse = preVoteResponse;
+        this.transportService = transportService;
+        this.startElection = startElection;
+
+        transportService.registerRequestHandler(REQUEST_PRE_VOTE_ACTION_NAME, Names.GENERIC, false, false,
+            PreVoteRequest::new,
+            (request, channel, task) -> channel.sendResponse(handlePreVoteRequest(request)));
+    }
+
+    /**
+     * Start a new pre-voting round.
+     * @param clusterState the last-accepted cluster state
+     * @param broadcastNodes the nodes from whom to request pre-votes
+     * @return the pre-voting round, which can be closed to end the round early.
+     */
+    public Releasable start(final ClusterState clusterState, final Iterable<DiscoveryNode> broadcastNodes) {
+        logger.debug("{} starting", this);
+        PreVotingRound currentRound = new PreVotingRound(clusterState, preVoteResponse.getCurrentTerm());
+        currentRound.start(broadcastNodes);
```

Comment:
not a fan of the "current"Round.

Reply:
Renamed in 4ec40e140c9caa1cf2cf73f562e4eb789c0749ec